### PR TITLE
Addition of Financial portal and settlement services

### PR DESCRIPTION
**Pull request by Kalaston**
This PR adds a new functionality to display the state of the financial platform:
a Finance portal that enables the interaction for multiple actions like transactions and settlement windows, which is a service of two containers: a back-end in Koa and a frontend in React:
Screenshot from 2019-12-10 04-14-19

This portal depends on the settlement service, which manages how the settlement windows are closed and how the settlements are saved, both actions being done in the _settlement management_ and _operator settlement_ respectively.

The portal can be accessed through a new NginX ingress that is being added in the mojaloop chart dependencies. This enables us to access the portal from outside the cluster.

### CODE REVIEW

File: finance-portal/values.yaml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+backend:
+  image:
+    repository: mojaloop/finance-portal-backend-service
+    tag: v0.0.1
+    pullPolicy: Always
+    port: 3000
+  init:
+    enabled: false
+    image:
+      repository: mojaloop/finance-portal-backend
```
Comment (lewisdaly) [inline finance-portal/values.yaml]: I think this should be `mojaloop/finance-portal-backend-service`, as the docker pull for the license scanner can't find it

Comment (vbarzokas) [reply finance-portal/values.yaml]: You are correct @lewisdaly, I updated it.

**Comment by vbarzokas** [comment]
> Looks good to me. However the license scans are failing still.
There is no information about why they fail in the logs though, do you have a clue?

**Comment by lewisdaly** [comment]
> > Looks good to me. However the license scans are failing still.
> There is no information about why they fail in the logs though, do you have a clue?

My only hint was `exit code 2`. Perhaps it's taking too long and circle is killing it? I've just re-run the step and we will see again.

File: finance-portal/templates/configmap-auth-init.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Values.config.is_populate.name | quote }}
+  labels:
+    app: {{ template "finance-portal.fullname" . }}
```
Comment (mdebarros) [inline finance-portal/templates/configmap-auth-init.yaml]: These labels are not following best practices --> https://helm.sh/docs/topics/chart_best_practices/labels/

See the following example: https://github.com/mojaloop/helm/blob/master/centralledger/chart-service/templates/deployment.yaml#L6

Please update.

File: finance-portal/templates/configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "finance-portal.fullname" . }}
+  labels:
+    app: {{ template "finance-portal.fullname" . }}
```
Comment (mdebarros) [inline finance-portal/templates/configmap.yaml]: Same comment about labels.

File: finance-portal/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,76 @@
+{{- $dbHost := ( .Values.db.host | replace "$release_name" .Release.Name ) -}}
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "finance-portal.fullname" . }}
+  labels:
```
Comment (mdebarros) [inline finance-portal/templates/deployment.yaml]: Same comment about labels apply here.

File: finance-portal/templates/ingress.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
```
Comment (mdebarros) [inline finance-portal/templates/ingress.yaml]: Please add labels, and align to the above label comments.

File: finance-portal/templates/ingress.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  annotations:
+    kubernetes.io/ingress.class: nginx
```
Comment (mdebarros) [inline finance-portal/templates/ingress.yaml]: Please move annotations to config values.yml.

Ingress template must be generic (and made specific in the values.yml via config). <-- i.e. it will allow any ingress-controller to be used, and thus it will not be specific to the nginx flavoured controller.

File: finance-portal/templates/ingress.yaml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    nginx.ingress.kubernetes.io/rewrite-target: /$1
+  name: {{ .Release.Name }}-ingress-service-portal-backend
+  namespace: {{ .Release.Namespace }}
+spec:
+  rules:
+  - http:
+      paths:
+      - path: /(.*)
+        backend:
+          serviceName: {{ .Release.Name }}-finance-portal
```
Comment (mdebarros) [inline finance-portal/templates/ingress.yaml]: Why only prefix is being derived from the release?

Please also derive the service name from the helm chart itself.

File: finance-portal/templates/service.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    - port: {{ .Values.backend.image.port }}
+      targetPort: backend
+      protocol: TCP
+      name: http-finance-portal-backend
+    - port: {{ .Values.frontend.image.port }}
+      targetPort: frontend
+      protocol: TCP
+      name: http-finance-portal-ui
+  selector:
+    app: {{ template "finance-portal.name" . }}
```
Comment (mdebarros) [inline finance-portal/templates/service.yaml]: Please update to reflect the updated labels.

File: mojaloop/values.yaml [hunk 9]
```diff
@@ -3483,6 +3485,9 @@ simulator:
 
   affinity: {}
 
+settlement:
+  enabled: true
```
Comment (mdebarros) [inline mojaloop/values.yaml]: Why was this added? I believe the settlement chart is already enabbled?

File: finance-portal/values.yaml [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    report1: "something"
+    report2: "something"
+
+frontend:
+  image:
+    repository: mojaloop/finance-portal-ui
+    tag: v0.0.3
+    pullPolicy: Always
+    port: 80
+  ingress:
```
Comment (mdebarros) [inline finance-portal/values.yaml]: This config is not used in the ingress definition. Please advise?

File: finance-portal/values.yaml [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    port: 3000
+  init:
+    enabled: false
+    image:
+      repository: mojaloop/finance-portal-backend-service
+      tag: v0.0.1
+      pullPolicy: Always
+  env: 'dev'
+  secretsMountPoint: '/secrets'
+  ingress:
```
Comment (mdebarros) [inline finance-portal/values.yaml]: This config is not used in the ingress definition. Please advise?

File: mojaloop/requirements.yaml [hunk 12]
```diff
@@ -32,3 +32,13 @@ dependencies:
   version: 8.7.1
   repository: "file://../transaction-requests-service"
   condition: transaction-requests-service.enabled
+- name: finance-portal
+  version: 8.6.0
+  repository: "file://../finance-portal"
+  condtion: finance-portal.enabled
+- name: nginx-ingress
```
Comment (mdebarros) [inline mojaloop/requirements.yaml]: Please remove the ingress controller.

This is a general k8s config, and should not be part of the mojaloop deployment.

It is up to the deployer on which Ingress controller they wish to utilise

File: mojaloop/requirements.yaml [hunk 13]
```diff
@@ -32,3 +32,13 @@ dependencies:
...
+  repository: https://kubernetes-charts.storage.googleapis.com
+  version: ~1.24.7
+- name: settlement
```
Comment (mdebarros) [inline mojaloop/requirements.yaml]: Please confirm why this was added?

It should already be included as part of central chart as the `central-settlement` helm chart.

File: mojaloop/values.yaml [hunk 14]
```diff
@@ -4477,3 +4482,8 @@ mojaloop-bulk:
           ## Used to pass Labels that are used by the Prometheus installed in your cluster to select Prometheus Rules to work with
           ## ref: https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#prometheusspec
     additionalLabels: {}
+
+nginx-ingress:
```
Comment (mdebarros) [inline mojaloop/values.yaml]: Please remove

File: settlement/Chart.yaml [hunk 15]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
```
Comment (mdebarros) [inline settlement/Chart.yaml]: Please confirm the purpose of this service, and how it differs from the central-settlement service?

File: settlement/templates/configmap.yaml [hunk 16]
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "settlement.fullname" . }}
+  labels:
+    app: {{ template "settlement.fullname" . }}
```
Comment (mdebarros) [inline settlement/templates/configmap.yaml]: Labels.

File: settlement/templates/deployment.yaml [hunk 17]
```diff
@@ -0,0 +1,85 @@
+{{- $dbHost := ( .Values.db.host | replace "$release_name" .Release.Name ) -}}
+
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "settlement.fullname" . }}
+  labels:
+    app: {{ template "settlement.name" . }}
```
Comment (mdebarros) [inline settlement/templates/deployment.yaml]: Labels.

File: settlement/templates/deployment.yaml [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    chart: {{ template "settlement.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  selector:
+    matchLabels:
+      app: {{ template "settlement.name" . }}
```
Comment (mdebarros) [inline settlement/templates/deployment.yaml]: Labels.

File: settlement/templates/deployment.yaml [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+      release: {{ .Release.Name }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "settlement.name" . }}
```
Comment (mdebarros) [inline settlement/templates/deployment.yaml]: Labels.

File: settlement/templates/ingress.yaml [hunk 20]
```diff
@@ -0,0 +1,67 @@
+{{- if .Values.operatorSettlement.ingress.enabled -}}
+{{- $serviceName := include "settlement.fullname" . -}}
+{{- $operatorSettlementPort := .Values.operatorSettlement.service.port -}}
+{{- $operatorSettlementPath := .Values.operatorSettlement.ingress.externalPath -}}
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ template "settlement.fullname" . }}
+  labels:
+    app: {{ template "settlement.fullname" . }}
```
Comment (mdebarros) [inline settlement/templates/ingress.yaml]: Labels.

File: settlement/templates/ingress.yaml [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+  {{- if .Values.operatorSettlement.ingress.tls }}
+  tls:
+    {{ toYaml .Values.operatorSettlement.ingress.tls | indent 4 }}
+  {{- end -}}
+{{- end -}}
+{{- if .Values.settlementManagement.ingress.enabled -}}
+{{- $serviceName := include "settlement.fullname" . -}}
+{{- $settlementManagementPort := .Values.settlementManagement.service.port -}}
+{{- $settlementManagementPath := .Values.settlementManagement.ingress.externalPath -}}
+---
```
Comment (mdebarros) [inline settlement/templates/ingress.yaml]: Please make ingress's names unique as I see 2 individual ingress resources defined here. 

Although they seem to be duplicated to some extent (i.e. they share the same annotations, etc)?

File: settlement/templates/ingress.yaml [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ template "settlement.fullname" . }}
+  labels:
+    app: {{ template "settlement.fullname" . }}
```
Comment (mdebarros) [inline settlement/templates/ingress.yaml]: Labels.

File: settlement/templates/secrets.yaml [hunk 23]
```diff
@@ -0,0 +1,20 @@
+{{- $dbHost := (.Values.db.host | replace "$release_name" .Release.Name) }}
+
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ template "settlement.fullname" . }}
+  labels:
+    app: {{ template "settlement.fullname" . }}
```
Comment (mdebarros) [inline settlement/templates/secrets.yaml]: Labels.

File: settlement/templates/service.yaml [hunk 24]
```diff
@@ -0,0 +1,23 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "settlement.fullname" . }}
+  labels:
+    app: {{ template "settlement.name" . }}
```
Comment (mdebarros) [inline settlement/templates/service.yaml]: Labels.

File: settlement/templates/service.yaml [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+    - port: {{ .Values.operatorSettlement.service.port }}
+      targetPort: operator
+      protocol: TCP
+      name: http-operator-settlement
+    - port: {{ .Values.settlementManagement.service.port }}
+      targetPort: management
+      protocol: TCP
+      name: http-settlement-management
+  selector:
+    app: {{ template "settlement.name" . }}
```
Comment (mdebarros) [inline settlement/templates/service.yaml]: Labels.

File: settlement/values.yaml [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+    enabled: true
+    externalPath: /
+    hosts:
+      api: settlement-management.local
+    annotations:
+      nginx.ingress.kubernetes.io/rewrite-target: '/'
+      # kubernetes.io/ingress.class: nginx
+      # kubernetes.io/tls-acme: "true"
+    tls:
+    # Secrets must be manually created in the namespace.
```
Comment (mdebarros) [inline settlement/values.yaml]: Please add a dummy secret. Refer to the example provided by the https://github.com/mojaloop/helm/blob/master/als-oracle-pathfinder/templates/secret.yaml

File: finance-portal/templates/deployment.yaml [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+    app: {{ template "finance-portal.name" . }}
+    chart: {{ template "finance-portal.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  selector:
+    matchLabels:
+      app: {{ template "finance-portal.name" . }}
```
Comment (mdebarros) [inline finance-portal/templates/deployment.yaml]: labels.

File: finance-portal/templates/deployment.yaml [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+      release: {{ .Release.Name }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "finance-portal.name" . }}
```
Comment (mdebarros) [inline finance-portal/templates/deployment.yaml]: labels.

File: central/values.yaml [hunk 29]
```diff
@@ -1413,6 +1413,7 @@ centralledger:
         wsrep_cluster_name=galera
         wsrep_sst_method=xtrabackup-v2
         pxc_strict_mode=DISABLED
+        sql_mode=STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION
```
[the remark this answers is not in the document]
Comment (msk-) [reply central/values.yaml]: 1. strictly speaking, they're not being added, they are in fact defaults: https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sql-mode-setting

2. You're right, I'm surprised that, unless I'm misreading this, they appear to be both defaults _and_ deprecated. Fun times. I suppose there's probably some backwards compat justification for that.

_Anyway_. This has arisen because the finance portal has some queries that (it turns out) aren't semantically correct, but previously mysql has been more forgiving (even in some 5.7 < 5.7.11 versions). In fact, another solution to this issue is to identify the affected queries and make them semantically correct. I think probably we shouldn't make the proposed change to the sql mode, and instead we should modify the queries in question.

[the remark this answers is not in the document]
Comment (msk-) [reply central/values.yaml]: Addressed by https://github.com/mojaloop/finance-portal-backend-service/pull/10

[the remark this answers is not in the document]
Comment (msk-) [reply central/values.yaml]: And https://github.com/mojaloop/finance-portal-backend-service/issues/11

File: mojaloop/values.yaml (hunk 9 above)
Comment (Kalaston) [reply mojaloop/values.yaml]: This is a new service, that handles settlements for the portal. I renamed the service into `portal-settlement`. Sorry for the confusion.

File: settlement/values.yaml (hunk 26 above)
Comment (Kalaston) [reply settlement/values.yaml]: There's already a secret in that service(https://github.com/mojaloop/helm/blob/finance-portal-service/portal-settlement/templates/secrets.yaml). I've removed the comment to reflect it.

Comment (mdebarros) [reply settlement/values.yaml]: > https://github.com/mojaloop/helm/blob/finance-portal-service/portal-settlement/templates/secrets.yaml

Yes but not for the TLS details.

File: portal-settlement/templates/secrets.yaml [hunk 30]
```diff
@@ -1,14 +1,14 @@
 {{- $dbHost := (.Values.db.host | replace "$release_name" .Release.Name) }}
-
 apiVersion: v1
 kind: Secret
 metadata:
-  name: {{ template "settlement.fullname" . }}
+  name: {{ template "portal-settlement.fullname" . }}
```
Comment (mdebarros) [inline portal-settlement/templates/secrets.yaml]: Would it be possible to use consistent naming with regard to portals?

I see that we have a `finance-portal`, then why not a `settlement-portal`?

File: portal-settlement/templates/operator-settlement-ingress.yaml [hunk 31]
```diff
@@ -0,0 +1,33 @@
+{{- if .Values.operatorSettlement.ingress.enabled -}}
+{{- $serviceName := include "portal-settlement.fullname" . -}}
+{{- $operatorSettlementPort := .Values.operatorSettlement.service.port -}}
+{{- $operatorSettlementPath := .Values.operatorSettlement.ingress.externalPath -}}
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ template "portal-settlement.fullname" . }}-operator-settlement
+  labels:
+    app: {{ template "portal-settlement.fullname" . }}
```
Comment (mdebarros) [inline portal-settlement/templates/operator-settlement-ingress.yaml]: Labels please.

File: portal-settlement/templates/settlement-management-ingress.yaml [hunk 32]
```diff
@@ -0,0 +1,33 @@
+{{- if .Values.settlementManagement.ingress.enabled -}}
+{{- $serviceName := include "portal-settlement.fullname" . -}}
+{{- $settlementManagementPort := .Values.settlementManagement.service.port -}}
+{{- $settlementManagementPath := .Values.settlementManagement.ingress.externalPath -}}
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ template "portal-settlement.fullname" . }}-settlement-management
+  labels:
+    app: {{ template "portal-settlement.fullname" . }}
```
Comment (mdebarros) [inline portal-settlement/templates/settlement-management-ingress.yaml]: Labels.

File: portal-settlement/templates/secrets.yaml (hunk 30 above)
Comment (Kalaston) [reply portal-settlement/templates/secrets.yaml]: although I agree with consistent naming, `settlement-portal` maybe sounds like another portal. the service is a complement for the finance-portal, so perhaps `finance-portal-settlement-management` is a better name.
Please let me know what you think.

File: finance-portal-settlement-management/templates/deployment.yaml [hunk 33]
```diff
@@ -0,0 +1,86 @@
...
+  labels:
+    app.kubernetes.io/name: {{ include "finance-portal-settlement-management.name" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/version: {{ .Chart.AppVersion }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+    helm.sh/chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+spec:
+  selector:
+    matchLabels:
+      app: {{ template "finance-portal-settlement-management.name" . }}
```
Comment (mdebarros) [inline finance-portal-settlement-management/templates/deployment.yaml]: label

File: finance-portal-settlement-management/templates/deployment.yaml [hunk 34]
```diff
@@ -0,0 +1,86 @@
...
+      release: {{ .Release.Name }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "finance-portal-settlement-management.name" . }}
```
Comment (mdebarros) [inline finance-portal-settlement-management/templates/deployment.yaml]: label

File: finance-portal-settlement-management/templates/service.yaml [hunk 35]
```diff
@@ -0,0 +1,24 @@
...
+    - port: {{ .Values.operatorSettlement.service.port }}
+      targetPort: operator
+      protocol: TCP
+      name: http-operator-settlement
+    - port: {{ .Values.settlementManagement.service.port }}
+      targetPort: management
+      protocol: TCP
+      name: http-settlement-management
+  selector:
+    app: {{ template "finance-portal-settlement-management.name" . }}
```
Comment (mdebarros) [inline finance-portal-settlement-management/templates/service.yaml]: label

File: finance-portal/templates/deployment.yaml (hunk 28 above)
Comment (mdebarros) [reply finance-portal/templates/deployment.yaml]: This has still not been updated?

File: finance-portal/templates/deployment.yaml (hunk 27 above)
Comment (mdebarros) [reply finance-portal/templates/deployment.yaml]: This has still not been updated?

File: finance-portal/templates/service.yaml (hunk 8 above)
Comment (mdebarros) [reply finance-portal/templates/service.yaml]: This has still not been updated?

File: finance-portal-settlement-management/Chart.yaml [hunk 36]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v1
+appVersion: "1.0"
+description: Portal Settlement
+name: finance-portal-settlement-management
+version: 8.6.0
+appVersion: "8.6.0"
```
Comment (mdebarros) [inline finance-portal-settlement-management/Chart.yaml]: duplicate `appVersion`

File: finance-portal-settlement-management/Chart.yaml [hunk 37]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v1
+appVersion: "1.0"
```
Comment (mdebarros) [inline finance-portal-settlement-management/Chart.yaml]: Please remove
